test/shared/src/main/scala/zio/test/Spec.scala

```diff
@@ -89,110 +106,162 @@ final case class Spec[+L, +T](caseValue: SpecCase[L, T, Spec[L, T]]) { self =>
    * reconstructing the spec with the same structure.
    */
-  final def foreach[R, E, A](f: T => ZIO[R, E, A]): ZIO[R, E, Spec[L, A]] =
-    foreachExec(ExecutionStrategy.Sequential)(f)
+  final def foreach[R1 <: R, E1, A](
+    failure: E => ZIO[R1, E1, A],
+    success: T => ZIO[R1, E1, A]
+  ): ZIO[R1, Nothing, Spec[R1, E1, L, A]] =
+    foreachExec(ExecutionStrategy.Sequential)(failure, success)
 
   /**
    * Iterates over the spec with the parallel strategy as the default, and
    * effectfully transforming every test with the provided function, finally
    * reconstructing the spec with the same structure.
    */
-  final def foreachPar[R, E, A](f: T => ZIO[R, E, A]): ZIO[R, E, Spec[L, A]] =
-    foreachExec(ExecutionStrategy.Parallel)(f)
+  final def foreachPar[R1 <: R, E1, A](
+    failure: E => ZIO[R1, E1, A],
+    success: T => ZIO[R1, E1, A]
+  ): ZIO[R1, Nothing, Spec[R1, E1, L, A]] =
+    foreachExec(ExecutionStrategy.Parallel)(failure, success)
 
   /**
    * Iterates over the spec with the parallel (`n`) strategy as the default, and
    * effectfully transforming every test with the provided function, finally
    * reconstructing the spec with the same structure.
    */
-  final def foreachParN[R, E, A](f: T => ZIO[R, E, A]): ZIO[R, E, Spec[L, A]] =
-    foreachExec(ExecutionStrategy.Parallel)(f)
+  final def foreachParN[R1 <: R, E1, A](
+    n: Int
+  )(failure: E => ZIO[R1, E1, A], success: T => ZIO[R1, E1, A]): ZIO[R1, Nothing, Spec[R1, E1, L, A]] =
+    foreachExec(ExecutionStrategy.ParallelN(n))(failure, success)
 
   /**
    * Returns a new spec with remapped labels.
    */
-  final def mapLabel[L1](f: L => L1): Spec[L1, T] =
-    transform[L1, T] {
+  final def mapLabel[L1](f: L => L1): Spec[R, E, L1, T] =
+    transform[R, E, L1, T] {
       case SuiteCase(label, specs, exec) => SuiteCase(f(label), specs, exec)
       case TestCase(label, test)         => TestCase(f(label), test)
     }
 
   /**
    * Returns a new spec with remapped tests.
    */
-  final def mapTest[T1](f: T => T1): Spec[L, T1] =
-    transform[L, T1] {
+  final def mapTest[T1](f: T => T1): Spec[R, E, L, T1] =
+    transform[R, E, L, T1] {
       case SuiteCase(label, specs, exec) => SuiteCase(label, specs, exec)
-      case TestCase(label, test)         => TestCase(label, f(test))
+      case TestCase(label, test)         => TestCase(label, test.map(f))
     }
 
   /**
-   * Computes the size of the spec, i.e. the number of tests in the spec.
+   * Uses the specified `Managed` to provide each test in this spec with its
+   * required environment.
    */
-  final def size: Int = fold[Int] {
-    case SuiteCase(_, counts, _) => counts.sum
-    case TestCase(_, _)          => 1
+  final def provideManaged[E1 >: E](managed: Managed[E1, R]): Spec[Any, E1, L, T] =
+    transform[Any, E1, L, T] {
+      case SuiteCase(label, specs, exec) => SuiteCase(label, specs.provideManaged(managed), exec)
```

Comment:
It seems like `transform` is awfully low-level for this, do we have anything that just allows function application to `specs` and `test`?

Reply:
I agree. How about something like:

```scala
  final def mapTests[R1, E1, L1 >: L, T1](
    suiteCase: ZIO[R, E, Vector[Spec[R1, E1, L1, T1]]] => ZIO[R1, E1, Vector[Spec[R1, E1, L1, T1]]],
    testCase: ZIO[R, E, T] => ZIO[R1, E1, T1]
  ): Spec[R1, E1, L1, T1] =
    caseValue match {
      case SuiteCase(label, specs, exec) =>
        Spec.suite(label, suiteCase(specs.map(_.map(_.mapTests(suiteCase, testCase)))), exec)
      case TestCase(label, test) => Spec.test(label, testCase(test))
    }
```

Then we could do:

```scala
  final def provideManaged[E1 >: E](managed: Managed[E1, R]): Spec[Any, E1, L, T] =
    mapTests[Any, E1, L, T](_.provideManaged(managed), _.provideManaged(managed))
```